src/main/java/usi/si/seart/service/MetricLanguageServiceImpl.java

```diff
@@ -0,0 +1,48 @@
+package usi.si.seart.service;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.util.ConcurrentReferenceHashMap;
+import usi.si.seart.model.MetricLanguage;
+import usi.si.seart.repository.MetricLanguageRepository;
+
+import javax.persistence.EntityManager;
+import javax.validation.constraints.NotNull;
+
+@Service
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+@AllArgsConstructor(onConstructor_ = @Autowired)
+public class MetricLanguageServiceImpl implements MetricLanguageService{
+
+    EntityManager entityManager;
+
+    ConcurrentReferenceHashMap<String, Object> metricLanguageLocks = new ConcurrentReferenceHashMap<>();
+
+    MetricLanguageRepository metricLanguageRepository;
+
+    private Object getMetricLanguageLock(String languageName) {
+        return this.metricLanguageLocks.compute(languageName, (k, v) -> v == null ? new Object() : v);
+    }
+
+    public MetricLanguage getOrCreateMetricLanguage(@NotNull String languageName) {
+        return metricLanguageRepository.findByLanguage(languageName).orElseGet(() -> {
+            // Acquires the lock for the metric language name
+            synchronized (getMetricLanguageLock(languageName)) {
+                // Checks whether the metric language entity has been created while awaiting the lock
+                MetricLanguage metricLang = metricLanguageRepository.findByLanguage(languageName).orElseGet(() ->
+                        // If not, creates it.
+                        metricLanguageRepository.save(
+                            MetricLanguage.builder()
+                                    .language(languageName)
+                                    .build()));
+                entityManager.clear();
+                entityManager.getEntityManagerFactory().getCache().evict(MetricLanguage.class, metricLang.getId());
```

Comment:
Instead of doing `entityManager.getEntityManagerFactory()`, I think you can just autowire it directly to the class. So instead of constantly running the getter, just add an `EntityManagerFactory` field. See [cleaning job implementation](https://github.com/seart-group/ghs/blob/develop/src/main/java/usi/si/seart/job/CleanUpProjectsJob.java).